src/note.rs

```diff
@@ -0,0 +1,394 @@
+use kelvin::{ByteHash, Content, Sink, Source};
+use std::convert::TryFrom;
+use std::convert::TryInto;
+use std::io::{self, Read, Write};
+
+use dusk_pki::jubjub_decode;
+use dusk_pki::Ownable;
+use dusk_pki::{PublicSpendKey, SecretSpendKey, StealthAddress, ViewKey};
+
+use dusk_plonk::jubjub::{dhke, GENERATOR_EXTENDED, GENERATOR_NUMS_EXTENDED};
+use poseidon252::cipher::PoseidonCipher;
+use poseidon252::sponge::sponge::sponge_hash;
+use rand::Rng;
+
+use crate::{
+    chunk_of, BlsScalar, Error, JubJubAffine, JubJubExtended, JubJubScalar,
+};
+
+const ENCRYPTED_DATA_SIZE: usize = 96;
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum NoteType {
+    Transparent = 0,
+    Obfuscated = 1,
+}
+
+impl TryFrom<u8> for NoteType {
+    type Error = Error;
+
+    fn try_from(note_type: u8) -> Result<Self, Self::Error> {
+        match note_type {
+            0 => Ok(NoteType::Transparent),
+            1 => Ok(NoteType::Obfuscated),
+            n => Err(Error::InvalidNoteType(n)),
+        }
+    }
+}
+
+impl TryFrom<i32> for NoteType {
+    type Error = Error;
+
+    fn try_from(note_type: i32) -> Result<Self, Self::Error> {
+        (note_type as u8).try_into()
+    }
+}
+
+/// A note that does not encrypt its value
+#[derive(Clone, Copy)]
+pub struct Note {
+    note_type: NoteType,
+    value_commitment: JubJubExtended,
+    nonce: JubJubScalar,
+    stealth_address: StealthAddress,
+    pos: u64,
+    encrypted_data: [u8; ENCRYPTED_DATA_SIZE],
```

Comment:
Why not use `PoseidonCipher` here? See: https://github.com/dusk-network/Poseidon252/blob/17329418740028b571752e936b190a1f15f94c5a/src/cipher/cipher.rs#L84

Reply:
First of all, because when this was implemented the first time we didn't have `PoseidonChiper` yet – that also why the `ECRYPTED_DATA_SIZE` is defined here. But once the `PoseidonChiper` was implemented, it was because the internals. It would be better if `PoseidonCipher` would use as internals bytes instead of `BlsScalar` – if we want to have this representation on contracts as well – but before changing either poseidon or phoenix-core it's better to cross that bridge and see what's the best approach.

Reply:
@khovratovich @vlopes11 any suggestions?

Reply:
We have on the same structure a `JubJubScalar`, so in this case we do not have only bytes. Why does this apply?